app/models/concerns/project_media_cached_fields.rb

```diff
@@ -715,6 +727,48 @@ def recalculate_tipline_search_results_count
       types = ["relevant_search_result_requests", "irrelevant_search_result_requests", "timeout_search_requests"]
       TiplineRequest.where(associated_type: 'ProjectMedia', associated_id: self.id, smooch_request_type: types).count
     end
+
+    def recalculate_media_cluster_origin(field = :origin) # Possible values for "field": :origin, :user_id, :timestamp
+      relationship = Relationship.where(target_id: self.id).last
+      origin = { origin: nil, user_id: nil, timestamp: nil }
+
+      # Not child of any media cluster
+      if relationship.nil?
+        if self.user == BotUser.smooch_user
+          origin[:origin] = CheckMediaClusterOrigins::OriginCodes::TIPLINE_SUBMITTED
+        else
+          origin[:origin] = CheckMediaClusterOrigins::OriginCodes::USER_ADDED
+        end
+        origin[:user_id] = self.user_id
+        origin[:timestamp] = self.created_at.to_i
+
+      # Child of a media cluster
+      else
+        if relationship.confirmed_at # A suggestion that was confirmed
+          origin[:origin] = CheckMediaClusterOrigins::OriginCodes::USER_MATCHED
+          origin[:user_id] = relationship.confirmed_by
+          origin[:timestamp] = relationship.confirmed_at.to_i
+        elsif relationship.user == BotUser.alegre_user
+          origin[:origin] = CheckMediaClusterOrigins::OriginCodes::AUTO_MATCHED
+          origin[:user_id] = relationship.user_id
+          origin[:timestamp] = relationship.created_at.to_i
+        elsif relationship.user.is_a?(User)
```

Comment:
I recommend using else without any conditions since the field is of User type. This way, we can move the user_id and timestamp fields outside the condition, as these fields are repeated in all cases.

Reply:
@melsawy , I can't replace the two `elsif`'s by a single `else` for now because we're currently not handling all cases yet (for example, not handling a relationship created the by Smooch Bot, between a caption and a media, for example). But that's a good point, I added a `FIXME` for now.